### PR TITLE
fix: Make containers wait for emulator-proxy to be healthy before starting (RQA-340)

### PR DESCRIPTION
**Pull request by DerekMaggio**
# Overview

Containers that depend on emulator-proxy now wait for it to be healthy before starting

# Changelog

- Update depends_on logic to include conditions
- Update tests

# Review requests

None

# Risk assessment

Low, it was already broken

### CODE REVIEW

File: emulation_system/emulation_system/compose_file_creator/conversion/service_builders/abstract_service_builder.py [hunk 1]
```diff
@@ -129,6 +130,17 @@ def _generate_container_name(
         )
         return container_name
 
+    def _cast_depends_on(
```
Comment (sfoster1) [inline emulation_system/emulation_system/compose_file_creator/conversion/service_builders/abstract_service_builder.py:133]: bit of a nit, but this really is not a cast in python - it actually changes values! It also is, apologies in advance, a monadic transform: it takes a container (an Optional) that contains a `None` or a `Some`, and it returns `Optional[None]` if it was given an `Optional[None]` and an `Optional[SomeOther]` if it was given an `Optional[Some]`. The best way to type this is with an overload:

```
# overloads define external interface
@typing.overload
def _cast_depends_on(self, depends_on: None) -> None: ...

@typing.overload
def _cast_depends_on(self, depends_on: IntermediateDependsOn) -> ServiceDependsOn: ...

# annotations on implementation allow typechecking of the function code
def _cast_depends_on(self, depends_on: Optional[IntermediateDependsOn]) -> Optional[ServiceDependsOn]:
    # code here
```

You can see the difference if you make a file like this:
```python
class A: pass

class B: pass

from typing import Optional, overload

def a(arg: Optional[A]) -> Optional[B]:
    if arg:
        return B()
    return None

@overload
def b(arg: A) -> B: ...

@overload
def b(arg: None) -> None: ...

def b(arg: Optional[A]) -> Optional[B]:
    if arg: return B()
    return None

reveal_type(a(None))
reveal_type(a(A()))
reveal_type(b(None))
reveal_type(b(A()))
```
```shell
mypy test.py 
test.py:22: note: Revealed type is "Union[test.B, None]"
test.py:23: note: Revealed type is "Union[test.B, None]"
test.py:24: note: Revealed type is "None"
test.py:25: note: Revealed type is "test.B"
Success: no issues found in 1 source file
```

for the function `b`, the return type can be inferred from the arg type.

**Review by sfoster1 — APPROVED** [review]
Looks good but you could do a really cool little typechecking thing here

**Review by fsinapi — APPROVED** [review]
I watched you run this and the behavior was correct so this LGTM!

File: emulation_system/emulation_system/compose_file_creator/conversion/service_builders/abstract_service_builder.py (hunk 1 above)
Comment (DerekMaggio) [reply emulation_system/emulation_system/compose_file_creator/conversion/service_builders/abstract_service_builder.py:136]: ae0ea725ba1951a0a38479b8129c7a69d7863828

**Pull request merged**